src/hub-connection.ts

```diff
@@ -1,78 +1,157 @@
+import { tap, map, filter, switchMap, skipUntil, debounceTime, take, delay } from "rxjs/operators";
 import { HubConnection as SignalRHubConnection } from "@aspnet/signalr-client";
 import { fromPromise } from "rxjs/observable/fromPromise";
 import { BehaviorSubject } from "rxjs/BehaviorSubject";
 import { Observable } from "rxjs/Observable";
+import { empty } from "rxjs/observable/empty";
+import { of } from "rxjs/observable/of";
 import { Observer } from "rxjs/Observer";
-import { tap } from "rxjs/operators";
 
 import { ConnectionState, ConnectionStatus, HubConnectionOptions } from "./hub-connection.model";
+import { buildQueryString } from "./utils/query-string";
+import { Dictionary } from "./utils/dictionary";
+import { emptyNext } from "./utils/rxjs";
 
 const connectedState: ConnectionState = { status: ConnectionStatus.connected };
+const connectionReadyState: ConnectionState = { status: ConnectionStatus.connectionReady };
```

Comment:
still not convinced with the name `connectionReady`.. perhaps  `prepareConnection` or `connecting` ... i thnk i like `connecting` more.. even tho its not entirely connecting with signalr just yet, but its still in the connecting phase